poc/opaque_core.sage

```diff
@@ -63,30 +63,64 @@ class OPAQUECore(object):
-        if self.config.mode == envelope_mode_custom_identifier:
-            cleartext_creds = CustomCleartextCredentials(response.pkS, creds.idU, creds.idS)
+    def recover_public_key(self, private_key):
+        sk = OS2IP(private_key)
+        if "ristretto" in self.config.group.name or "decaf" in self.config.group.name:
+            sk = OS2IP_le(private_key)
+        pk = sk * self.config.group.generator()
+        return self.config.group.serialize(pk)
+
+    def derive_group_key_pair(self, seed):
+        skS = self.config.group.hash_to_scalar(seed, dst=_as_bytes("OPAQUE-HashToScalar"))
+        pkS = skS * self.config.group.generator()
+        return (skS, pkS)
+
+    def build_inner_envelope(self, random_pwd, envelope_nonce, client_private_key):
+        if self.config.mode == envelope_mode_internal:
+            Nsk = self.config.Nsk
+            seed = self.config.kdf.expand(random_pwd, envelope_nonce + _as_bytes("PrivateKey"), Nsk)
+            (_, client_public_key) = self.derive_group_key_pair(seed)
+            pk_bytes = self.config.group.serialize(client_public_key)
+            return (InnerEnvelope(), self.config.group.serialize(client_public_key))
+        if self.config.mode == envelope_mode_external:
+            secret_creds = SecretCredentials(client_private_key)
+            serialized_creds = secret_creds.serialize()
+            pseudorandom_pad = self.config.kdf.expand(random_pwd, envelope_nonce + _as_bytes("Pad"), len(serialized_creds))
+            encrypted_creds = xor(serialized_creds, pseudorandom_pad)
+            client_public_key = self.recover_public_key(client_private_key)
+            return (InnerEnvelope(encrypted_creds), client_public_key)
+        raise Exception("Unsupported mode")
+
+    def create_cleartext_credentials(self, server_public_key, client_public_key, server_identity, client_identity):
+        if server_identity == None:
+            server_identity = server_public_key
+        if client_identity == None:
+            client_identity = client_public_key
+        return CleartextCredentials(server_public_key, client_identity, server_identity)
+
+    def create_envelope(self, creds, random_pwd, server_public_key):
+        envelope_nonce = random_bytes(OPAQUE_NONCE_LENGTH)
+        Nh = self.config.hash().digest_size
+        auth_key = self.config.kdf.expand(random_pwd, envelope_nonce + _as_bytes("AuthKey"), Nh)
+        export_key = self.config.kdf.expand(random_pwd, envelope_nonce + _as_bytes("ExportKey"), Nh)
+        masking_key = self.derive_masking_key(random_pwd)
 
-        pt = secret_creds.serialize()
-        auth_data = cleartext_creds.serialize()
+        inner_env, client_public_key = self.build_inner_envelope(random_pwd, envelope_nonce, creds.skU)
+        cleartext_creds = self.create_cleartext_credentials(server_public_key, client_public_key, creds.idS, creds.idU)
+        auth_tag = self.config.mac.mac(auth_key, inner_env.serialize() + cleartext_creds.serialize())
```

Comment:
auth_tag needs to be computed over the mode and the nonce as well, not just the inner_env and cleartext_creds

Reply:
Agreed, that's where the vectors don't match :)